In [1]:
from __future__ import print_function
import torch
import torch.nn as nn
import torch.nn.functional as F
import torch.optim as optim
from torchvision import datasets, transforms

In [2]:
# Train Phase transformations
train_transforms = transforms.Compose([
                                       transforms.Resize((28, 28)), #augmentation
                                       transforms.ColorJitter(brightness=0.10, contrast=0.1, saturation=0.10, hue=0.1), #augmentation
                                       transforms.ToTensor(),
                                       transforms.Normalize((0.1307,), (0.3081,)) # The mean and std have to be sequences (e.g., tuples), therefore you should add a comma after the values.
                                       # Note the difference between (0.1307) and (0.1307,)
                                       ])

# Test Phase transformations
test_transforms = transforms.Compose([
                                       transforms.ToTensor(),
                                       transforms.Normalize((0.1307,), (0.3081,))
                                       ])


In [3]:
train = datasets.MNIST('./data', train=True, download=True, transform=train_transforms)
test = datasets.MNIST('./data', train=False, download=True, transform=test_transforms)

Failed to download (trying next):
HTTP Error 403: Forbidden



100%|██████████| 9.91M/9.91M [00:00<00:00, 18.0MB/s]


Extracting ./data/MNIST/raw/train-images-idx3-ubyte.gz to ./data/MNIST/raw

Failed to download (trying next):
HTTP Error 403: Forbidden



100%|██████████| 28.9k/28.9k [00:00<00:00, 498kB/s]


Extracting ./data/MNIST/raw/train-labels-idx1-ubyte.gz to ./data/MNIST/raw

Failed to download (trying next):
HTTP Error 403: Forbidden



100%|██████████| 1.65M/1.65M [00:00<00:00, 4.51MB/s]


Extracting ./data/MNIST/raw/t10k-images-idx3-ubyte.gz to ./data/MNIST/raw

Failed to download (trying next):
HTTP Error 403: Forbidden



100%|██████████| 4.54k/4.54k [00:00<00:00, 3.75MB/s]

Extracting ./data/MNIST/raw/t10k-labels-idx1-ubyte.gz to ./data/MNIST/raw



In [4]:
SEED = 1

# CUDA?
cuda = torch.cuda.is_available()
print("CUDA Available?", cuda)

# For reproducibility
torch.manual_seed(SEED)

if cuda:
    torch.cuda.manual_seed(SEED)

# dataloader arguments - something you'll fetch these from cmdprmt
dataloader_args = dict(shuffle=True, batch_size=128, num_workers=4, pin_memory=True) if cuda else dict(shuffle=True, batch_size=64)

# train dataloader
train_loader = torch.utils.data.DataLoader(train, **dataloader_args)

# test dataloader
test_loader = torch.utils.data.DataLoader(test, **dataloader_args)

CUDA Available? True


/usr/local/lib/python3.10/dist-packages/torch/utils/data/dataloader.py:617: UserWarning: This DataLoader will create 4 worker processes in total. Our suggested max number of worker in current system is 2, which is smaller than what this DataLoader is going to create. Please be aware that excessive worker creation might get DataLoader running slow or even freeze, lower the worker number to avoid potential slowness/freeze if necessary.
  warnings.warn(


In [5]:
class Model2(nn.Module):
    def __init__(self):
        super(Model2, self).__init__()

        self.convblock1 = nn.Sequential(
            nn.Conv2d(in_channels=1, out_channels=10, kernel_size=(3, 3), padding=0),
            nn.BatchNorm2d(10),
            nn.ReLU()
        ) # INPUT 28*28 | OUTPUT 26*26 | RF 3

        self.convblock2 = nn.Sequential(
            nn.Conv2d(in_channels=10, out_channels=10, kernel_size=(3, 3), padding=0),
            nn.BatchNorm2d(10),
            nn.ReLU()
        ) # INPUT 26*26 | OUTPUT 24*24 | RF 5

        self.convblock3 = nn.Sequential(
            nn.Conv2d(in_channels=10, out_channels=10, kernel_size=(3, 3), padding=0),
            nn.BatchNorm2d(10),
            nn.ReLU()
        ) # INPUT 24*24 | OUTPUT 22*22 | RF 7

        self.pool1 = nn.MaxPool2d(2, 2) # INPUT 22*22 | OUTPUT 11*11 | RF 8

        self.convblock4 = nn.Sequential(
            nn.Conv2d(in_channels=10, out_channels=10, kernel_size=(3, 3), padding=0),
            nn.BatchNorm2d(10),
            nn.ReLU()
        ) # INPUT 11*11 | OUTPUT 9*9 | RF 12

        self.convblock5 = nn.Sequential(
            nn.Conv2d(in_channels=10, out_channels=16, kernel_size=(3, 3), padding=0),
            nn.BatchNorm2d(16),
            nn.ReLU()
        ) # INPUT 9*9 | OUTPUT 7*7 | RF 16


        self.convblock6 = nn.Sequential(
            nn.Conv2d(in_channels=16, out_channels=10, kernel_size=(3, 3), padding=0),
            nn.BatchNorm2d(10),
            nn.ReLU()
        ) # INPUT 7*7 | OUTPUT 5*5 | RF 20

        self.convblock7 = nn.Sequential(
            nn.Conv2d(in_channels=10, out_channels=10, kernel_size=(3, 3), padding=0)
        ) # INPUT 5*5 | OUTPUT 3*3 | RF 24

        self.gap = nn.Sequential(
            nn.AvgPool2d(3)
        ) # INPUT 3*3 | OUTPUT 1*1 | RF 28

        self.dropout = nn.Dropout(0.1) #increased dropout


    def forward(self, x):
        x = self.convblock1(x)
        x = self.convblock2(x)
        x = self.dropout(x)  #dropout
        x = self.convblock3(x)
        x = self.pool1(x)
        x = self.convblock4(x)
        x = self.convblock5(x)
        x = self.dropout(x) #dropout
        x = self.convblock6(x)
        x = self.convblock7(x)
        x = self.gap(x)
        x = x.view(-1, 10)
        return F.log_softmax(x, dim=-1)

In [6]:
!pip install torchsummary
from torchsummary import summary
use_cuda = torch.cuda.is_available()
device = torch.device("cuda" if use_cuda else "cpu")
print(device)
model = Model2().to(device)
summary(model, input_size=(1, 28, 28))

cuda
----------------------------------------------------------------
        Layer (type)               Output Shape         Param #
            Conv2d-1           [-1, 10, 26, 26]             100
       BatchNorm2d-2           [-1, 10, 26, 26]              20
              ReLU-3           [-1, 10, 26, 26]               0
            Conv2d-4           [-1, 10, 24, 24]             910
       BatchNorm2d-5           [-1, 10, 24, 24]              20
              ReLU-6           [-1, 10, 24, 24]               0
           Dropout-7           [-1, 10, 24, 24]               0
            Conv2d-8           [-1, 10, 22, 22]             910
       BatchNorm2d-9           [-1, 10, 22, 22]              20
             ReLU-10           [-1, 10, 22, 22]               0
        MaxPool2d-11           [-1, 10, 11, 11]               0
           Conv2d-12             [-1, 10, 9, 9]             910
      BatchNorm2d-13             [-1, 10, 9, 9]              20
             ReLU-14             [

In [7]:
from tqdm import tqdm

train_losses = []
test_losses = []
train_acc = []
test_acc = []

def train(model, device, train_loader, optimizer, epoch):
  model.train()
  pbar = tqdm(train_loader)
  correct = 0
  processed = 0
  for batch_idx, (data, target) in enumerate(pbar):
    # get samples
    data, target = data.to(device), target.to(device)

    # Init
    optimizer.zero_grad()
    # In PyTorch, we need to set the gradients to zero before starting to do backpropragation because PyTorch accumulates the gradients on subsequent backward passes.
    # Because of this, when you start your training loop, ideally you should zero out the gradients so that you do the parameter update correctly.

    # Predict
    y_pred = model(data)

    # Calculate loss
    loss = F.nll_loss(y_pred, target)
    train_losses.append(loss)

    # Backpropagation
    loss.backward()
    optimizer.step()

    # Update pbar-tqdm

    pred = y_pred.argmax(dim=1, keepdim=True)  # get the index of the max log-probability
    correct += pred.eq(target.view_as(pred)).sum().item()
    processed += len(data)

    pbar.set_description(desc= f'Loss={loss.item()} Batch_id={batch_idx} Accuracy={100*correct/processed:0.2f}')
    train_acc.append(100*correct/processed)

def test(model, device, test_loader):
    model.eval()
    test_loss = 0
    correct = 0
    with torch.no_grad():
        for data, target in test_loader:
            data, target = data.to(device), target.to(device)
            output = model(data)
            test_loss += F.nll_loss(output, target, reduction='sum').item()  # sum up batch loss
            pred = output.argmax(dim=1, keepdim=True)  # get the index of the max log-probability
            correct += pred.eq(target.view_as(pred)).sum().item()

    test_loss /= len(test_loader.dataset)
    test_losses.append(test_loss)

    print('\nTest set: Average loss: {:.4f}, Accuracy: {}/{} ({:.2f}%)\n'.format(
        test_loss, correct, len(test_loader.dataset),
        100. * correct / len(test_loader.dataset)))

    test_acc.append(100. * correct / len(test_loader.dataset))

In [8]:
model =  Model2().to(device)
optimizer = optim.SGD(model.parameters(), lr=0.01, momentum=0.9)
EPOCHS = 15
for epoch in range(EPOCHS):
    print("EPOCH:", epoch)
    train(model, device, train_loader, optimizer, epoch)
    test(model, device, test_loader)

EPOCH: 0


Loss=0.07724960893392563 Batch_id=468 Accuracy=91.64: 100%|██████████| 469/469 [00:32<00:00, 14.60it/s]



Test set: Average loss: 0.0671, Accuracy: 9799/10000 (97.99%)

EPOCH: 1


Loss=0.026071548461914062 Batch_id=468 Accuracy=98.02: 100%|██████████| 469/469 [00:36<00:00, 12.97it/s]



Test set: Average loss: 0.0478, Accuracy: 9858/10000 (98.58%)

EPOCH: 2


Loss=0.08313081413507462 Batch_id=468 Accuracy=98.41: 100%|██████████| 469/469 [00:32<00:00, 14.52it/s]



Test set: Average loss: 0.0479, Accuracy: 9828/10000 (98.28%)

EPOCH: 3


Loss=0.04827054217457771 Batch_id=468 Accuracy=98.66: 100%|██████████| 469/469 [00:31<00:00, 14.86it/s]



Test set: Average loss: 0.0415, Accuracy: 9870/10000 (98.70%)

EPOCH: 4


Loss=0.024755120277404785 Batch_id=468 Accuracy=98.69: 100%|██████████| 469/469 [00:32<00:00, 14.36it/s]



Test set: Average loss: 0.0331, Accuracy: 9893/10000 (98.93%)

EPOCH: 5


Loss=0.010424304753541946 Batch_id=468 Accuracy=98.91: 100%|██████████| 469/469 [00:31<00:00, 14.93it/s]



Test set: Average loss: 0.0312, Accuracy: 9908/10000 (99.08%)

EPOCH: 6


Loss=0.03885653242468834 Batch_id=468 Accuracy=98.96: 100%|██████████| 469/469 [00:31<00:00, 15.11it/s]



Test set: Average loss: 0.0315, Accuracy: 9896/10000 (98.96%)

EPOCH: 7


Loss=0.009951157495379448 Batch_id=468 Accuracy=99.01: 100%|██████████| 469/469 [00:31<00:00, 14.99it/s]



Test set: Average loss: 0.0294, Accuracy: 9914/10000 (99.14%)

EPOCH: 8


Loss=0.12311068922281265 Batch_id=468 Accuracy=99.06: 100%|██████████| 469/469 [00:31<00:00, 14.97it/s]



Test set: Average loss: 0.0247, Accuracy: 9919/10000 (99.19%)

EPOCH: 9


Loss=0.055809687823057175 Batch_id=468 Accuracy=99.08: 100%|██████████| 469/469 [00:31<00:00, 14.75it/s]



Test set: Average loss: 0.0240, Accuracy: 9925/10000 (99.25%)

EPOCH: 10


Loss=0.038612861186265945 Batch_id=468 Accuracy=99.14: 100%|██████████| 469/469 [00:31<00:00, 14.68it/s]



Test set: Average loss: 0.0230, Accuracy: 9919/10000 (99.19%)

EPOCH: 11


Loss=0.009483086876571178 Batch_id=468 Accuracy=99.20: 100%|██████████| 469/469 [00:33<00:00, 13.95it/s]



Test set: Average loss: 0.0258, Accuracy: 9919/10000 (99.19%)

EPOCH: 12


Loss=0.015231919474899769 Batch_id=468 Accuracy=99.19: 100%|██████████| 469/469 [00:31<00:00, 15.09it/s]



Test set: Average loss: 0.0251, Accuracy: 9922/10000 (99.22%)

EPOCH: 13


Loss=0.013160977512598038 Batch_id=468 Accuracy=99.25: 100%|██████████| 469/469 [00:31<00:00, 14.99it/s]



Test set: Average loss: 0.0287, Accuracy: 9917/10000 (99.17%)

EPOCH: 14


Loss=0.005055147223174572 Batch_id=468 Accuracy=99.25: 100%|██████████| 469/469 [00:32<00:00, 14.47it/s]



Test set: Average loss: 0.0215, Accuracy: 9927/10000 (99.27%)



**Target:**

Reduce the parameter count

Keep the epochs within 15

Reach test accuracy greater than 99.4

**Results:**

Parameters: 6,778

Best Training Accuracy: 99.25

Best Test Accuracy: 99.27

**Analysis:**

Parameter count has been reduced

The model has been significantly optimized, achieving a much lighter architecture.

Target accuracy has not been reached

**What can be done**

Introduce ReduceLROnPlateau scheduler for dynamically updating the learning rate

Try using Adam optimizer

Gradually increase the number of feature maps in intermediate layers but still keep the count under 8000
